### PR TITLE
timestamp: support multiple trailer timestamp formats

### PR DESCRIPTION
**Pull request by wdebruij**
Make packet trailer timestamp format configurable.

EF_RX_TIMESTAMPING_ORDERING allows selecting timestamps embedded in packet trailers for ordering and SO_TIMESTAMPING. The current implementation assumes a single fixed trailer format, cPacket.

Make the format selectable, with EF_RX_TIMESTAMPING_TRAILER_FORMAT. Introduce initially two new formats to demonstrate the feature.

Implementation:

- expand ci_rx_pkt_timestamp_cpacket into ci_rx_pkt_timestamp_trailer, to demultiplex type. Minimal refactoring of the existing function to avoid code duplication. Cpacket func can be simplified further, but with more churn, so left for a v2 revision.
- expand src/tests/unit/header/ci/internal/ip_timestamp.c with basic unit tests for the new types.

For more information on cpacket and ttag trailer formats, also see
- https://github.com/wireshark/wireshark/blob/master/epan/dissectors/packet-metamako.c
- https://github.com/wireshark/wireshark/blob/master/epan/dissectors/packet-cisco-ttag.c

Tested: scripts/run_unit_tests.sh

### CODE REVIEW

File: src/include/ci/internal/ip_timestamp.h [hunk 1]
```diff
@@ -165,14 +164,91 @@ ci_rx_pkt_timestamp_cpacket(const ci_ip_pkt_fmt* pkt,
...
+
+  const struct ttag* ttag = (const struct ttag*)buf_end - 1;
+
+  uint64_t ts = ((uint64_t)ttag->ts[0] << 40) | ((uint64_t)ttag->ts[1] << 32) |
+                ((uint64_t)ttag->ts[2] << 24) | ((uint64_t)ttag->ts[3] << 16) |
+                ((uint64_t)ttag->ts[4] << 8)  | ((uint64_t)ttag->ts[5]);
+
+  const uint64_t nsec_per_sec = 1000ULL * 1000 * 1000;
+  ts_out->sec = ts / nsec_per_sec;
+  ts_out->nsec = ts - (ts_out->sec * nsec_per_sec);
```
Comment (mseymour-xilinx) [inline src/include/ci/internal/ip_timestamp.h:185]:
```suggestion
  ts_out->nsec = ts - (ts_out->sec * nsec_per_sec);
  ts_out->nsec_frac = 0;
```

File: src/include/ci/internal/ip_timestamp.h [hunk 2]
```diff
@@ -165,14 +164,91 @@ ci_rx_pkt_timestamp_cpacket(const ci_ip_pkt_fmt* pkt,
...
+}
```
Comment (mseymour-xilinx) [inline src/include/ci/internal/ip_timestamp.h:186]:
```suggestion
  ts_out->nsec_frac = 0;
}
```

File: src/tests/unit/header/ci/internal/ip_timestamp.c [hunk 3]
```diff
@@ -442,6 +482,52 @@ static void test_rx_pkt_timestamp_cpacket_pcap_wireshark(void)
...
+  pkt = &buf->pkt;
+  pkt_init(pkt);
+  pkt_udp(pkt, NULL, 0);
+  pkt_ttag(pkt, sec, nsec);
+  STATE_STASH(buf);
+
+  ci_rx_pkt_timestamp_trailer(pkt, ts, CITP_RX_TIMESTAMPING_TRAILER_FORMAT_TTAG);
+  STATE_CHECK(ts, sec, sec);
+  STATE_CHECK(ts, nsec, nsec);
+
```
Comment (mseymour-xilinx) [inline src/tests/unit/header/ci/internal/ip_timestamp.c:503]:
```suggestion
  STATE_CHECK(ts, nsec_frac, 0);

```

**Review by mseymour-xilinx — APPROVED** [review]
Looks good. It would be nice to refactor `ci_rx_pk_timestamp_cpacket` as described.

File: src/include/ci/internal/ip_timestamp.h (hunk 2 above)
Comment (wdebruij) [reply src/include/ci/internal/ip_timestamp.h:181]: Thanks. Done

**Comment by wdebruij** [comment]
Thanks for the review! I pushed a new versions with the suggestions incorporated. And a few other changes. Added a changelog to the bottom of the commit message (below the ---)

**Comment by wdebruij** [comment]
Just a note that I pushed a [commit ](https://github.com/wdebruij/kerneltools/commit/d6e2aa7b090593ce54a82664d720fd910b06930f) to my own repo two tools with which I tested this end-to-end.

- [trailer_tx](https://github.com/wdebruij/kerneltools/blob/master/tests/trailer_tx.c) is a small packet socket program to send an UDP packet, optionally with a trailer timestamp in one of the three formats. This does not require Onload.
- [trailer_rx](https://github.com/wdebruij/kerneltools/blob/master/tests/trailer_rx.c) exercises onload_timestamping_request to read both the NIC and trailer timestamp (if present). This duplicates some test already in Onload, but the current format works particularly well with trailer_tx.

**Review by abower-amd — APPROVED** [review]

**Comment by abower-amd** [comment]
Thanks for the contribution, @wdebruij !

**Comment by wdebruij** [comment]
Thanks for accepting it!

**Pull request merged**